package.json

```diff
@@ -1,7 +1,7 @@
 {
```

Comment:
To me, the (confusingly named) json-schema module represents the types that our system supports, so if someone wants to work on contributors, they can use the `Contributor` type, and similarly for references and `BibliographyItem`. I don't see a problem of being json-schema dependent; the whole point of the module is that.

Whether or not it's appropriate to use encode/decode here, I would like for the time being to keep ProseMirror-related logic contained within body-editor, and any external communication should happen through json-schema, so a component in the style-guide should not work with PM.

> We should never encode models/nodes individually. There is no assurance that encoding and decoding process will not affect any other nodes or these nodes will not be affected.

Why is there no guarantee? Do you know of any case where we actually update the nodes? If so, we should remove that logic, I can't think of any case where we would want the input node to be updated.

> It feels like this approach is a bit undermined by not converting author modals to actually rely on prosemirror document, which, as I commented on LEAN-3083 PR, is something what needs to be done to transfer the modals here. What happens here is that we move authors modals out of the article-editor but they are still completely json-schema dependent, but now instead of going through the main process of preparing data in the article-editor we process nodes here separately, which makes things more spread out and less clear and puts a contributors node view in charge of a very responsible process of data conversion.

I disagree with that, using the store for data communication is way more confusing and more spread out. Currently if I wanted to see what data is being used in the modal, I would have to:
- Go the node view to find out the component is actually passed in props.
- Go to the article-editor to find what the component being passed is.
- Go to the component to find out where the data is coming from.
- Go to the store init logic to find out how this data is managed.

Reply:
I also think adapt-tracked-data should be in body-editor, where it can also be used in there.

Reply:
> To me, the (confusingly named) json-schema module represents the types that our system supports, so if someone wants to work on contributors, they can use the `Contributor` type, and similarly for references and `BibliographyItem`. I don't see a problem of being json-schema dependent; the whole point of the module is that.

To clarify, when I wrote about JSON-schema I mean data set and not the types - that is the model map that we get by decoding the document. The problem with this is that for almost every json-schema based type we have a peer type in the prosemirror document to use both no the same level is going to be confusing, would it not?


> Why is there no guarantee? Do you know of any case where we actually update the nodes? If so, we should remove that logic, I can't think of any case where we would want the input node to be updated.

It was never designed to work that way: some nodes may reach to other nodes for additional info. For example ref ids (all rid), section priorities are dependent on the general document context. It would be perfectly acceptable if a node would need to reach out to its peer or parent during encoding or decoding, we also have some nodes for which there are no peer json-schema models at all. 



> I disagree with that, using the store for data communication is way more confusing and more spread out. Currently if I wanted to see what data is being used in the modal, I would have to:
> 
> * Go the node view to find out the component is actually passed in props.
> * Go to the article-editor to find what the component being passed is.
> * Go to the component to find out where the data is coming from.
> * Go to the store init logic to find out how this data is managed.

Perhaps it may be somewhat confusing but it is definitely not spread out, since data handling happens in a single place.
- Not sure why we need to pass any components. We don't pass any components as things are now. 
- Not clear why you need to go the component to find out where the data are coming from if they always come only from trackedModelMap. It is rather in this change where things will become less clear as it creates an adhoc case and you would need go searching where exactly data are created in this specific case.
- If by store init logic you mean the complicated approach of building author and affiliations data, then I agree it's something what could've been improved but it doesn't need the modal to be rendered and managed through a node view to be improved.


> Whether or not it's appropriate to use encode/decode here, I would like for the time being to keep ProseMirror-related logic contained within body-editor, and any external communication should happen through json-schema, so a component in the style-guide should not work with PM.

While I agree with that we need to contain prosemirror related logic to the body-editor, even if we move adaptTrackData here, still, we will have to import it to the article-editor, because operations on trackedModelMap will require it and as I see it, it should be used only there. Also article-editor will still have to deal with the doc and it has a lot of prosemirror related logic besides adaptTrackedData: in inspector, in snapshots handling, and in track changes. It also receives the PM doc from api and initiates the body-editor, basically acting as a controller for the body-editor. So while it is possible to abstract json-schema away from the body-editor it doesn't seem likely for the article-editor to be unaware of prosemirror doc and its parts.

In my opinion, If we prefer to revise how we manage data relations between the article-editor and body-editor, we should improve on that and take advantage of that improvement on the application level rather than dealing with this on case by case basis.

Reply:
> It was never designed to work that way: some nodes may reach to other nodes for additional info. For example ref ids (all rid), section priorities are dependent on the general document context. It would be perfectly acceptable if a node would need to reach out to its peer or parent during encoding or decoding, we also have some nodes for which there are no peer json-schema models at all.

That's fair. You're saying that `encode` might need data from somewhere else in the document, not that `encode` might update the nodes.

> While I agree with that we need to contain prosemirror related logic to the body-editor, even if we move adaptTrackData here, still, we will have to import it to the article-editor, because operations on trackedModelMap will require it and as I see it, it should be used only there. Also article-editor will still have to deal with the doc and it has a lot of prosemirror related logic besides adaptTrackedData: in inspector, in snapshots handling, and in track changes. It also receives the PM doc from api and initiates the body-editor, basically acting as a controller for the body-editor. So while it is possible to abstract json-schema away from the body-editor it doesn't seem likely for the article-editor to be unaware of prosemirror doc and its parts.

Yeah, article-editor is a different story which is not what I'm trying to change here, and I don't see a reason why we should change how it works. I do think we should move the adapt logic to body-editor, but it will still be used in article-editor.

----

So is the only problem you have here the usage of  `encode` on individual nodes? If so that can be changed.

Reply:
"That's fair. You're saying that encode might need data from somewhere else in the document, not that encode might update the nodes."
- that's right